### PR TITLE
feat!: Matches required query parameters

### LINKED ISSUE #20: Support Query Parameters?
Would IPFS be interested in supporting `_redirects` files with _query_ parameters, similar to [Netlify](https://docs.netlify.com/routing/redirects/#syntax-for-the-redirects-file:~:text=for%20more%20details.-,query,-%3A%20Query%20string%20parameters))? (Re: ipfs/kubo#8890). If this seems valuable, I'd put a PR together (here and in ipfs/kubo).

I would find it valuable to be able to write a redirects file similar to the one below, to be able to provide responses to 'dynamic-looking' web-requests (from clients that expect to use specific, query parameter-based, GET APIs) via IPFS:

```plain _redirects
/things type=:type /things/:type.html 200
/things /things.html 200
```

(This is identical to the format [Netlify uses](https://docs.netlify.com/routing/redirects/redirect-options/#query-parameters) — only _required_ parameters must be specified.)

`GET /things?type=photos` → Served from `/things/photos.html`
`GET /things` → Served from `/things.html`

### PR DESCRIPTION
**Pull request by jphastings**
Code now `Parse`s and `MatchAndExpandPlaceholders` for query parameters as well as paths. _I ended up implementing #20 without waiting for a discussion on the issue, for which I apologise! I'll be happy to rewrite without a review if there are concerns with the proposal._

Placeholders are shared between the two.

```plain
/things type=photos /photos.html 200
/things type=       /invalid-thing-type.html 400
/things type=:thing /thing-:thing.html 200
/things             /things.html 200
```

This is a **breaking change** for the package's API (though not for the `_redirects` file format), as the query params need to be passed into `MatchAndExpandPlaceholders`. I'll be un-drafting ipfs/boxo#406 (which adapts syntax to include this work), once (if!) this PR is merged, so `go.mod`/`go.sum` can be corrected.

This commit also codifies in tests some of the previously implicit edge cases, particularly around duplicate placeholders.

Closes #20

---

If you build a copy of `ipfs` with this in (via ipfs/boxo#406) then you can visit [this CID locally](http://bafybeiegqaac7fcpfdvy35d73nwxnmtu27kbnt4rnyu7pghhff5wzfuagq.ipfs.localhost:8080) and see this in action.

### CODE REVIEW

File: redirects_test.go [hunk 1]
```diff
@@ -108,7 +130,12 @@ func FuzzParse(f *testing.F) {
 		"/%C4%85 /ę 301\n",
 		"#/a \n\n/b",
 		"/a200 /b200 200\n/a301 /b301 301\n/a302 /b302 302\n/a303 /b303 303\n/a307 /b307 307\n/a308 /b308 308\n/a404 /b404 404\n/a410 /b410 410\n/a451 /b451 451\n",
-		"hello\n", "/redirect-one /one.html\r\n/200-index /index.html 200\r\n", "a b 2\nc   d 42", "/a/*/b blah", "/from https://example.com 200\n/a/:blah/yeah /b/:blah/yeah"}
+		"hello\n", "/redirect-one /one.html\r\n/200-index /index.html 200\r\n", "a b 2\nc   d 42", "/a/*/b blah", "/from https://example.com 200\n/a/:blah/yeah /b/:blah/yeah",
+		"/fixed-val val=val /to\n", "/dynamic-val val=:val /to/:val\n", "/empty-val val= /to\n", "/any-val val /to\n",
+		"/fixed-val val=val /to 200\n/dynamic-val val=:val /to/:val 301\n/empty-val val= /to 404\n/any-val val /to 302\n",
+		"/multi-query val1=val1 val2=:val2 val3= val4 /to/:val2\n/multi-query2 val1=val1 val2=:val2 val3= val4 /to/:val2 302\n",
+		"/bad-syntax1 val=a&val=b /to\n", "/bad-syntax2 val=a&val2=b /to 302\n", "/a ^&notparams /b\n", "/bad-status type=:type /to 3oo\n", "/bad-chars :type=whatever /to\n", "/bad-chars type=what:ever /to\n",
+	}
```
Comment (jphastings) [inline redirects_test.go:138]: I wasn't 100% clear on the fuzz testing strategy here, I'd appreciate a closer look at whether I've chosen suitable testcases here.

**Comment by lidel** [comment]
(turning to a draft until we have an IPIP mentioned in https://github.com/ipfs/go-ipfs-redirects-file/issues/20)

**Comment by lidel** [comment]
Closing as the IPIP did not happen + we went with simpler support – details in https://github.com/ipfs/go-ipfs-redirects-file/issues/20#issuecomment-2737288073
